src/stm32l0.c

```diff
@@ -0,0 +1,1076 @@
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_PEKEYR(nvm),  STM32Lx_NVM_PEKEY1);
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_PEKEYR(nvm),  STM32Lx_NVM_PEKEY2);
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_OPTKEYR(nvm), STM32Lx_NVM_OPTKEY1);
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_OPTKEYR(nvm), STM32Lx_NVM_OPTKEY2);
+
+  return !(adiv5_ap_mem_read (ap, STM32Lx_NVM_PECR(nvm))
+           & STM32Lx_NVM_PECR_OPTLOCK);
+}
+
+
+/** Erase a region of flash using a stub function.  This only works
+    when the MCU hasn't entered a fault state (see NOTES).  The flash
+    array is erased for all pages from addr to addr+len inclusive. */
+static int stm32lx_nvm_prog_erase_stubbed (struct target_s* target,
+                                           uint32_t addr, int size)
+{
+  struct stm32lx_nvm_stub_info info;
+  const uint32_t nvm = stm32lx_nvm_phys (target);
+
+  info.nvm       = nvm;
+  info.page_size = stm32lx_nvm_prog_page_size (target);
+
+  /* Load the stub */
+  target_mem_write_words (target, STM32Lx_STUB_PHYS,
+                          (void*) &stm32l0_nvm_prog_erase_stub[0],
+                          sizeof (stm32l0_nvm_prog_erase_stub));
+
+  /* Setup parameters */
+  info.destination = addr;
+  info.size        = size;
+
+  /* Copy parameters */
+  target_mem_write_words (target, STM32Lx_STUB_INFO_PHYS,
+                          (void*) &info, sizeof (info));
+
+  /* Execute stub */
+  target_pc_write (target, STM32Lx_STUB_PHYS);
+  if (target_check_error (target))
+    return -1;
+  target_halt_resume (target, 0);
+  while (!target_halt_wait (target))
+    ;
+  {
+    ADIv5_AP_t* ap = adiv5_target_ap(target);
+    if (adiv5_ap_mem_read (ap, STM32Lx_NVM_SR(nvm)) & STM32Lx_NVM_SR_ERR_M)
+      return -1;
+  }
+
+  return 0;
+}
+
+
+/** Write to program flash using a stub function.  This only works
+    when the MCU hasn't entered a fault state.  Once the MCU faults,
+    this function will not succeed because the MCU will fault before
+    executing a single instruction in the stub. */
+static int stm32lx_nvm_prog_write_stubbed (struct target_s* target,
+                                           uint32_t destination,
+                                           const uint8_t* source,
+                                           int size)
+{
+  struct stm32lx_nvm_stub_info info;
+  const uint32_t nvm = stm32lx_nvm_phys (target);
+  const size_t page_size = stm32lx_nvm_prog_page_size (target);
+
+  /* We can't handle unaligned destination or non-word writes. */
+  /* *** FIXME: we should handle misaligned writes by padding with
+     *** zeros.  Probably, the only real time we'd see something
+     *** misaligned would be on a write to a final half-word.  Perhaps
+     *** this could be handled with the stub?  In fact, aligning the
+     *** start is going to be mandatory.  We will change the code to
+     *** cope with a trailing half-word. */
+  if ((destination & 3) || (size & 3))
+    return -1;
```

Comment:
On 3/7/15 7:06 PM, Gareth McMullin wrote:
>
> In src/stm32l0.c 
> <https://github.com/blacksphere/blackmagic/pull/75#discussion_r26001878>:
>
> > +                                           const uint8_t* source,
> > +                                           int size)
> > +{
> > +  struct stm32lx_nvm_stub_info info;
> > +  const uint32_t nvm = stm32lx_nvm_phys (target);
> > +  const size_t page_size = stm32lx_nvm_prog_page_size (target);
> > +
> > +  /* We can't handle unaligned destination or non-word writes. */
> > +  /* *** FIXME: we should handle misaligned writes by padding with
> > +     *** zeros.  Probably, the only real time we'd see something
> > +     *** misaligned would be on a write to a final half-word.  Perhaps
> > +     *** this could be handled with the stub?  In fact, aligning the
> > +     *** start is going to be mandatory.  We will change the code to
> > +     *** cope with a trailing half-word. */
> > +  if ((destination & 3) || (size & 3))
> > +    return -1;
>
> Although rare, cases do exist where the section bounds are unaligned. 
> See how this is handled in the STM32F1 driver.
>
>


OK.  I'll look at the fix there and make the necessary change.